components/percona_telemetry/data_provider.cc

```diff
@@ -0,0 +1,317 @@
+}
+
+bool DataProvider::collect_plugins_info(rapidjson::Document *document) {
+  QueryResult result;
+  if (do_query("SELECT PLUGIN_NAME FROM information_schema.plugins WHERE "
+               "PLUGIN_STATUS='ACTIVE'",
+               &result)) {
+    return true;
+  }
+
+  rapidjson::Document::AllocatorType &allocator = document->GetAllocator();
+
+  rapidjson::Value plugins(rapidjson::Type::kArrayType);
+
+  for (auto &plugin_iter : result) {
+    rapidjson::Value plugin_name;
+    plugin_name.SetString(plugin_iter[0].c_str(), allocator);
+    plugins.PushBack(plugin_name, allocator);
+  }
+  document->AddMember(rapidjson::StringRef("active_plugins"), plugins,
+                      allocator);
+
+  return false;
+}
+
+bool DataProvider::collect_components_info(rapidjson::Document *document) {
+  QueryResult result;
+  if (do_query("SELECT component_urn FROM mysql.component", &result)) {
+    return true;
+  }
+
+  rapidjson::Document::AllocatorType &allocator = document->GetAllocator();
+
+  rapidjson::Value components(rapidjson::Type::kArrayType);
+
+  for (auto &component_iter : result) {
+    rapidjson::Value component_name;
+    component_name.SetString(component_iter[0].c_str(), allocator);
+    components.PushBack(component_name, allocator);
+  }
+  document->AddMember(rapidjson::StringRef("active_components"), components,
+                      allocator);
+  return false;
+}
+
+bool DataProvider::collect_uptime_info(rapidjson::Document *document) {
+  QueryResult result;
+  if (do_query("SHOW GLOBAL STATUS LIKE 'Uptime'", &result)) {
+    return true;
+  }
+
+  rapidjson::Document::AllocatorType &allocator = document->GetAllocator();
+  rapidjson::Value uptime;
+  uptime.SetString(result[0][1].c_str(), allocator);
+  document->AddMember(rapidjson::StringRef("uptime"), uptime, allocator);
+  return false;
+}
+
+bool DataProvider::collect_dbs_number_info(rapidjson::Document *document) {
+  QueryResult result;
+  if (do_query(
+          "SELECT COUNT(*) FROM information_schema.SCHEMATA WHERE SCHEMA_NAME "
+          "NOT IN('mysql', 'information_schema', 'performance_schema', 'sys')",
+          &result)) {
+    return true;
+  }
+
+  rapidjson::Document::AllocatorType &allocator = document->GetAllocator();
+  rapidjson::Value db_cnt;
+  db_cnt.SetString(result[0][0].c_str(), allocator);
+  document->AddMember(rapidjson::StringRef("databases_count"), db_cnt,
+                      allocator);
+  return false;
+}
+
+/* Note that this metric is update very X, so it may be inacurate.
+   We could make it accurate but that would need ANALYZE TABLE for every table
+   which would be overkill.*/
+bool DataProvider::collect_dbs_size_info(rapidjson::Document *document) {
+  QueryResult result;
+  if (do_query("SELECT IFNULL(ROUND(SUM(data_length + index_length), 1), '0') "
+               "size_MB FROM information_schema.tables WHERE table_schema NOT "
+               "IN('mysql', 'information_schema', 'performance_schema', 'sys')",
```

Comment:
There are 2 places. If we extract it to the constant, we will need to construct the query separately. And that will lead to the conclusion that we need to move the whole query to the constant and replace the code with the constant, like (`do_query(collect_dbs_info_query)`). But the query itself is coupled with parsing of the query result, so in my opinion, it will not make it any easier/cleaner.